### ISSUE TITLE
Failure to infer implicit ABC (`Iterable` and `enumerate`)

### ISSUE THREAD

**Issue by dr-carlos** [astral-sh/ty#2768]
### Summary

Hi,

I've noticed that `ty` seems to not automatically infer that a class implements an ABC if it's not explicitly specified, at least in this specific case (https://play.ty.dev/edd5db26-937f-445a-876d-e921f2641d31):

```python
from typing import Iterable, reveal_type, Self, TYPE_CHECKING

if TYPE_CHECKING:
    class A[T]:
        def __init__(self, iterable: Iterable[T]) -> None: ...
        def __iter__(self) -> Self: ...
        def __next__(self) -> T: ...

def func(a: A[str]):
    reveal_type(a) # A[str]
    enumerated = enumerate(a)
    reveal_type(enumerated) # enumerate[Unknown]; should be enumerate[str]
    for i, s in enumerated:
        reveal_type(i) # int
        reveal_type(s) # Unknown; should be str
```

If `class A[T]` specifies inheritance from `Iterable[T]` (which it already implictly satisfies) - then `ty` infers correctly (https://play.ty.dev/8dd033ab-7aa7-4b7f-ab6c-ba7d3bfde41a):

```python
from typing import Iterable, reveal_type, Self, TYPE_CHECKING

if TYPE_CHECKING:
    class A[T](Iterable[T]):
        def __init__(self, iterable: Iterable[T]) -> None: ...
        def __iter__(self) -> Self: ...
        def __next__(self) -> T: ...

def func(a: A[str]):
    reveal_type(a) # A[str]
    enumerated = enumerate(a)
    reveal_type(enumerated) # enumerate[str]
    for i, s in enumerated:
        reveal_type(i) # int
        reveal_type(s) # str
```

This comes up in the real world with classes like `itertools.islice` - which isn't explicitly declared as `Iterable`: https://github.com/python/typeshed/blob/main/stdlib/itertools.pyi

`mypy` correctly identifies the types in both cases.

```python
test.py:11: note: Revealed type is "test.A[builtins.str]"
test.py:13: note: Revealed type is "builtins.enumerate[builtins.str]"
test.py:15: note: Revealed type is "builtins.int"
test.py:16: note: Revealed type is "builtins.str"
```

Apologies if there's already an open issue for this - I couldn't find any specific matches in the issue search.

### Version

ty 0.0.15

**Comment by AlexWaygood** [comment]
Thanks! Please see #1714
